main/lsp/requests/completion.cc

```diff
@@ -172,26 +244,39 @@ LSPResult LSPLoop::handleTextDocumentCompletion(unique_ptr<core::GlobalState> gs
         auto resp = move(queryResponses[0]);
 
         if (auto sendResp = resp->isSend()) {
-            auto pattern = sendResp->callerSideName.data(*gs)->shortName(*gs);
-            auto receiverType = sendResp->dispatchResult->main.receiver;
-            logger->debug("Looking for method similar to {}", pattern);
-            auto methodsMap = findSimilarMethodsIn(*gs, receiverType, pattern);
-            auto methods = vector<pair<core::NameRef, vector<core::SymbolRef>>>(methodsMap.begin(), methodsMap.end());
-            fast_sort(methods, [&](auto leftPair, auto rightPair) -> bool {
-                auto leftShortName = leftPair.first.data(*gs)->shortName(*gs);
-                auto rightShortName = rightPair.first.data(*gs)->shortName(*gs);
+            auto prefix = sendResp->callerSideName.data(*gs)->shortName(*gs);
+            logger->debug("Looking for method similar to {}", prefix);
+
+            auto similarMethods = allSimilarMethods(*gs, *sendResp->dispatchResult, prefix);
+            fast_sort(similarMethods, [&](const auto &left, const auto &right) -> bool {
+                if (left.depth != right.depth) {
+                    return left.depth < right.depth;
+                }
+
+                auto leftShortName = left.method.data(*gs)->name.data(*gs)->shortName(*gs);
+                auto rightShortName = right.method.data(*gs)->name.data(*gs)->shortName(*gs);
                 if (leftShortName != rightShortName) {
                     return leftShortName < rightShortName;
                 }
-                return leftPair.first._id < rightPair.first._id;
+
+                return left.method._id < right.method._id;
             });
-            for (auto &[methodName, methodSymbols] : methods) {
-                if (methodSymbols[0].exists()) {
-                    fast_sort(methodSymbols, [&](auto lhs, auto rhs) -> bool { return lhs._id < rhs._id; });
-                    items.push_back(getCompletionItem(*gs, methodSymbols[0], receiverType,
-                                                      sendResp->dispatchResult->main.constr.get()));
+
+            auto deduped = vector<SimilarMethod>{};
+            auto lastName = core::NameRef::noName();
+            for (auto &similarMethod : similarMethods) {
+                auto name = originalName(*gs, similarMethod.method.data(*gs)->name);
+
+                if (lastName != name) {
+                    deduped.emplace_back(similarMethod);
```

Comment:
This is only deduping identical entries with the same `depth`, right? So it wouldn't dedupe those at different depths.

Reply:
Good catch. Given that I don't really know what I want the heuristic to look like for completion items, I'm going to add a test to document this current behavior, and leave it in.

Reply:
Totally fair. I raised in case you were not aware, and since I lack some of the sorbet knowledge you have that may inform you of situations where this might occur (abstract methods?).